File: backend/Models/states/GameState.cs
using System.Text.Json;
using backend.Models.entities;
using backend.Models.enums;

namespace backend.Models.states;

public class GameState
{
    #region Properties
    public GameStatus GameStatus { get; set; }
    public int TurnNumber { get; set; }
    public Guid RoomId { get; set; }
    public List<Player> PlayersList { get; set; }
    public List<List<HexTile>> HexTilesList { get; set; }
    
    #endregion

    
    public string Serialize()
    {
        return JsonSerializer.Serialize(this);
    }

    public GameState Deserialize(string data)
    {
       return JsonSerializer.Deserialize<GameState>(data)!;
    }

    public GameState(GameStatus gameStatus, List<Player> playersList, List<List<HexTile>> hexTilesList, int turnNumber, Guid roomId)
    {
        GameStatus = gameStatus;
        PlayersList = playersList;
        HexTilesList = hexTilesList;
        TurnNumber = turnNumber;
        RoomId = roomId;
    }
}

File: frontend/civ_hex/lib/models/gamestate.dart
import 'package:civ_hex/components/hex_tile.dart';
import 'package:civ_hex/enums/game_status.dart';
import 'package:civ_hex/models/player.dart';

class GameState {
  GameStatus gameStatus;
  num turnNumber;
  String roomId;
  List<Player> playersList;
  List<List<HexTile>> hexTilesList;

  GameState(
      {required this.gameStatus,
      required this.turnNumber,
      required this.roomId,
      required this.playersList,
      required this.hexTilesList});

  factory GameState.fromJson(Map<String, dynamic> json) {
    return GameState(
      gameStatus: GameStatus.values[json['GameStatus']],
      turnNumber: json['TurnNumber'],
      roomId: json['RoomId'],
      playersList: (json['PlayersList'] as List).map((i) => Player.fromJson(i)).toList(),
      hexTilesList: (json['HexTilesList'] as List).map((i) => (i as List).map((j) => HexTile.fromJson(j)).toList()).toList(),
    );
  }
}
